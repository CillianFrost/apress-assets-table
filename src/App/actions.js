export const LOAD_TABLE_DATA_START = 'LOAD_TABLE_DATA_START';

export const loadTableDataStart = (id, payload) => ({
  type: LOAD_TABLE_DATA_START,
  payload
});
