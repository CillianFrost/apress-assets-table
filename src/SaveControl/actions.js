export const SAVE_START = 'SAVE_START';
export const SAVE_SUCCESS = 'SAVE_SUCCESS';
export const SAVE_CREATE_DIFF = 'SAVE_CREATE_DIFF';
export const SAVE_DIFF = 'SAVE_DIFF';
export const CONTINUE_SAVE = 'CONTINUE_SAVE';
export const SUCCESS_REMOVE_MESSAGE = 'SUCCESS_REMOVE_MESSAGE';


export const saveStart = () => ({
  type: SAVE_START,
});

export const saveSuccess = (payload = {}) => ({
  type: SAVE_SUCCESS,
  payload,
});

export const saveCreateDiff = (payload) => ({
  type: SAVE_CREATE_DIFF,
  payload,
});

export const continueSave = (payload) => ({
  type: CONTINUE_SAVE,
  payload,
});
