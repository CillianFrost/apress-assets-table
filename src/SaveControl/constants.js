const successRemoveMessageDelay = 3000;

export default successRemoveMessageDelay;
