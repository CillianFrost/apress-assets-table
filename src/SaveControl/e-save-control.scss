@import '../styles/mixins/icons';

.e-save-control {
  color: #1f2532; font-size: 12px; line-height: 60px; display: inline-block; margin: 0 0 0 30px;
  &.is-success:before { margin: 0 4px -2px 0; display: inline-block; @include check-icon(); width: 12px; height: 12px; content: ''; }
  &.is-progress:before { @include spin-loader-icon(); margin: 0 4px 0 0; display: inline-block; width: 12px; height: 12px; content: ''; }
  &.is-error:before { @include close-icon(); margin: 0 4px -2px 0; display: inline-block; width: 12px; height: 12px; content: ''; }
}
