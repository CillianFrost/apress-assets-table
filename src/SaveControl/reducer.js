import {
  SAVE_SUCCESS,
  SAVE_REPEAT,
  SAVE_CREATE_DIFF,
  SAVE_START,
  SAVE_DIFF
} from './actions';
import {
  TABLE_EDITOR_LOAD_SUCCESS,
  TABLE_EDITOR_SET_TEXT,
  TABLE_EDITOR_CELL_END_DRAG,
  TABLE_EDITOR_ROW_ADD_DEFAULT_ID,
  TABLE_EDITOR_ROW_ADD_ID,
  TABLE_EDITOR_SET_IMAGES,
  TABLE_EDITOR_CELL_END_DRAG_IMAGES,
  TABLE_EDITOR_ROW_ADD,
} from '../Table/actions';

const initialState = {
  isSave: false,
  isError: false,
  prevState: [],
  isProgress: false,
  fetchDiff: false,
  waitingState: [],
  saveState: []
};

export default function (state = initialState, action) {
  switch (action.type) {
    case TABLE_EDITOR_LOAD_SUCCESS:
      return {
        ...state,
        prevState: [...action.payload.rows]
      };

    case SAVE_REPEAT:
    case TABLE_EDITOR_CELL_END_DRAG:
    case TABLE_EDITOR_CELL_END_DRAG_IMAGES:
    case TABLE_EDITOR_SET_TEXT:
    case TABLE_EDITOR_SET_IMAGES:
    case TABLE_EDITOR_ROW_ADD:
    case 'HISTORY_PREV':
    case 'HISTORY_NEXT':
      return {
        ...state,
        isSave: true
      };

    case SAVE_CREATE_DIFF:
      return {
        ...state,
        fetchDiff: true,
        isSave: false,
      };

    case SAVE_DIFF:
      return {
        ...state,
        waitingState: action.payload.waitingState,
        fetchDiff: false,
        prevState: action.payload.prevState
      };

    case SAVE_START:
      return {
        ...state,
        saveState: state.waitingState,
        isProgress: true,
        isError: false,
        waitingState: []
      };

    case TABLE_EDITOR_ROW_ADD_DEFAULT_ID:
    case TABLE_EDITOR_ROW_ADD_ID: {
      const tmpPrevState = state.prevState.map((row) => {
        const payloadItem = action.payload.find(payloadRow =>
          row.check.common.id === payloadRow.id);

        if (payloadItem) {
          return {
            ...row,
            check: {
              ...row.check,
              common: {
                ...row.check.common,
                id: payloadItem.record_id
              }
            }
          };
        }

        return row;
      });

      const tmpWaitingState = state.waitingState.map((row) => {
        const payloadItem = action.payload.find(payloadRow => row.id === payloadRow.id);

        if (payloadItem) {
          return {
            ...row,
            id: payloadItem.record_id
          };
        }

        return row;
      });

      return {
        ...state,
        waitingState: tmpWaitingState,
        prevState: tmpPrevState
      };
    }

    case SAVE_SUCCESS: {
      if (!action.payload.error) {
        return {
          ...state,
          saveState: [],
          isProgress: false,
        };
      }

      const saveState = state.saveState;

      state.waitingState.forEach((record) => {
        const saveItem = saveState.find(item => item.id === record.id);

        if (saveItem) {
          if (saveItem.destroy && !record.destroy) {
            delete saveItem.destroy;
          }

          Object.keys(record).forEach((key) => {
            if (typeof record[key] === 'object' && record[key] !== null) {
              saveItem[key] = {...saveItem[key], ...record[key]};
            } else {
              saveItem[key] = record[key];
            }
          });
        } else {
          saveState.push(record);
        }
      });

      return {
        ...state,
        saveState: [],
        waitingState: saveState,
        isProgress: false,
        isSave: true,
        isError: true
      };
    }

    default:
      return state;
  }
}
