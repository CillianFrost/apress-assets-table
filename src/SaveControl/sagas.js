import {put, call, select} from 'redux-saga/effects';
import {delay} from 'redux-saga';
import _isEqual from 'lodash/isEqual';
import {api} from '../utils';
import {TREE_LOAD_START} from '../Tree/actions';
import {TABLE_EDITOR_ROW_ADD_ID, TABLE_EDITOR_ROW_ADD_DEFAULT_ID} from '../Table/actions';
import {ERROR_REMOVE} from '../Error/actions';
import {SAVE_SUCCESS, SAVE_DIFF} from './actions';

let newId = -100000;

const putSave = rows =>
  api.put(app.config.urlSaveTiger, {rows})
    .then(response => response.data);

const pollingJob = jobId =>
  api.get(`${app.config.urlJob}/${jobId}`)
    .then(response => response.data);

const getSave = props => props.save;

const getFieldId = () => 'check';

const find = (secondRow, row) =>
  (row.id && secondRow.id && row.id === secondRow.id) ||
  (row.check && secondRow.check &&
  row.check.common.id === secondRow.check.common.id);

const transformForServer = records =>
  records.map((record) => {
    const newObj = {
      id: record.check.common.id,
      columns: {}
    };

    Object.keys(record).forEach((key) => {
      if (key !== 'check') {
        if (record[key].common) {
          // кастомное сохранение для поля с фотографиями
          if (record[key].common.images) {
            // привязка массива с фото
            if (record[key].common.images[0].src === '') {
              newObj.columns[key] = {images: []};
            } else {
              newObj.columns[key] =
                {images: record[key].common.images.map(image => ({id: image.id}))};
            }
          } else {
            newObj.columns[key] = record[key].common;
          }
        } else {
          newObj[key] = record[key];
        }
      }
    });

    return newObj;
  });

const validate = (column, row) => {
  if (row[column] && typeof row[column] === 'object' && row[column] !== null) {
    switch (column) {
      case 'name':
        return !!row.name.common.text;

      default:
        return true;
    }
  }

  return true;
};

const clearEmptyFields = (obj) => {
  if (typeof obj === 'object' && obj !== null) {
    const newObj = {};
    Object.keys(obj).forEach((key) => {
      const childField = clearEmptyFields(obj[key]);

      if (childField && Object.keys(childField).length) {
        newObj[key] = childField;
      }

      if (childField && typeof childField !== 'object') {
        newObj[key] = childField;
      }
    });

    return newObj;
  }

  return obj;
};

const getDeletedItems = (cur, prev) => {
  const deletedItems = [];
  prev.forEach((prevChild) => {
    const curChild = cur.find(find.bind({}, prevChild));

    if (!curChild) {
      deletedItems.push({
        check: {
          common: {
            id: prevChild.check.common.id
          }
        },
        destroy: true
      });
    }
  });

  return deletedItems;
};

const getNewItems = (cur, prev) => {
  const newItems = [];
  cur.forEach((curChild) => {
    const prevChild = prev.find(find.bind({}, curChild));

    if (!prevChild && validate('name', curChild)) {
      newItems.push(clearEmptyFields(curChild));
    }
  });

  return newItems;
};

const getDiff = (cur, prev) => {
  if (Array.isArray(prev) && ((cur.length && cur[0].src) || (prev.length && prev[0].src))) {
    if (!(_isEqual(cur, prev))) {
      if (!cur.length) {
        return [{src: ''}];
      }
      return cur;
    }
  }

  if (Array.isArray(prev) && prev.length) {
    const tmpDiff = [];
    prev.forEach((prevChild) => {
      const curChild = cur.find(find.bind({}, prevChild));
      if (curChild) {
        const childDiff = getDiff(curChild, prevChild);

        if (childDiff && typeof childDiff === 'object' && Object.keys(childDiff).length) {
          if (Object.keys(prevChild).includes(getFieldId())) {
            childDiff[getFieldId()] = prevChild[getFieldId()];
          }

          tmpDiff.push(childDiff);
        }

        if (childDiff && typeof childDiff !== 'object') {
          tmpDiff.push(childDiff);
        }
      }
    });
    return tmpDiff;
  }

  if (typeof prev === 'object' && prev !== null) {
    const objDiff = {};
    Object.keys(prev).forEach((childKey) => {
      if (validate(childKey, cur)) {
        const childDiff = getDiff(cur[childKey], prev[childKey]);

        if (childDiff && Object.keys(childDiff).length) {
          objDiff[childKey] = childDiff;
        }

        if ((childDiff && typeof childDiff !== 'object') || childDiff === null) {
          objDiff[childKey] = childDiff;
        }
      }
    });

    return objDiff;
  }

  return cur !== prev ? (cur || null) : undefined;
};

const createDiff = (curState, prevState) => {
  const diff = getDiff(curState, prevState);
  const deletedItems = getDeletedItems(curState, prevState);
  const newItems = getNewItems(curState, prevState);

  return transformForServer([...diff, ...deletedItems, ...newItems]);
};

export function* saveCreateDiff(action) {
  try {
    const diff = createDiff(action.payload.curState, action.payload.prevState);
    const state = yield select(getSave);
    let waitingState = state.waitingState;
    if (waitingState.length) {
      diff.forEach((record) => {
        const waitingItem = waitingState.find(item => item.id === record.id);

        if (waitingItem) {
          if (waitingItem.destroy && !record.destroy) {
            delete waitingItem.destroy;
          }

          Object.keys(record).forEach((key) => {
            if (typeof record[key] === 'object' && record[key] !== null) {
              waitingItem[key] = {...waitingItem[key], ...record[key]};
            } else {
              waitingItem[key] = record[key];
            }
          });
        } else {
          waitingState.push(record);
        }
      });
    } else {
      waitingState = diff;
    }

    yield put({
      type: SAVE_DIFF,
      payload: {
        waitingState,
        prevState: action.payload.curState
      }
    });
  } catch (err) {
    console.log(err);
  }
}

export function* save() {
  try {
    yield put({type: ERROR_REMOVE, payload: {target: 'save'}});

    const saveProps = yield select(getSave);

    if (saveProps.saveState.length) {
      const job = yield call(putSave.bind({}, saveProps.saveState));
      let response = {};

      while (true) {
        yield call(delay, 1000);
        response = yield call(pollingJob.bind({}, job.job_id));

        if (response.failed || response.succeeded) {
          break;
        }
      }

      if (response.succeeded) {
        const deletedItems = saveProps.saveState.filter(row => row.destroy);
        if (deletedItems.length) {
          yield put({
            type: TABLE_EDITOR_ROW_ADD_DEFAULT_ID,
            payload: deletedItems.map((item) => {
              const data = {
                id: item.id,
                record_id: newId
              };
              newId -= 1;

              return data;
            })
          });
        }

        yield put({type: TABLE_EDITOR_ROW_ADD_ID, payload: response.payload});
        yield put({type: TREE_LOAD_START, payload: null});
      }
    }

    yield put({type: SAVE_SUCCESS, payload: {error: false}});
  } catch (err) {
    yield put({type: SAVE_SUCCESS, payload: {error: true}});
  }
}
