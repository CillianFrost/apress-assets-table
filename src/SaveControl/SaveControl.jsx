import React from 'react';
import PropTypes from 'prop-types';
import {block} from '../utils';
import './e-save-control.scss';

const b = block('e-save-control');

const getMessage = (message, isError, isProgress) => {
  if (isError) {
    return message.error;
  }

  if (isProgress) {
    return message.progress;
  }

  return message.success;
};

const SaveControl = props =>
  <div
    className={b.is({
      success: !props.isError && !props.isProgress,
      error: props.isError,
      progress: props.isProgress || props.removeInProgrees
    })}
  >
    {getMessage(props.message, props.isError, props.isProgress)}
  </div>;

SaveControl.propTypes = {
  isProgress: PropTypes.bool,
  isError: PropTypes.bool,
  removeInProgrees: PropTypes.bool,
  message: PropTypes.shape({
    error: PropTypes.string,
    progress: PropTypes.string,
    success: PropTypes.string
  })
};

export default SaveControl;
