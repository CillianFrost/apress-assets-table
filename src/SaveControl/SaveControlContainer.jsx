import React, {Component} from 'react';
import {connect} from 'react-redux';
import SaveControl from './SaveControl';

class SaveControlContainer extends Component {
  componentWillReceiveProps(nextProps) {
    const {isSave, fetchDiff, isProgress, waitingState, prevState} = nextProps.save;
    const {rows: curState} = nextProps;
    const {removeInProgrees} = this.props;

    if (isSave && !fetchDiff) {
      nextProps.actions.saveCreateDiff({curState, prevState});
    }

    if (!isProgress && !fetchDiff && waitingState.length) {
      nextProps.actions.saveStart();
    }

    if (isProgress || waitingState.length || fetchDiff || removeInProgrees) {
      window.addEventListener('beforeunload', this.hendlerBeforeunload);
    } else {
      window.removeEventListener('beforeunload', this.hendlerBeforeunload);
    }
  }

  hendlerBeforeunload = (e) => {
    const message = 'Возможно внесенные изменения не сохранятся';

    if (e) { e.returnValue = message; }

    return message;
  }

  render() {
    return (
      <SaveControl
        message={this.props.message}
        removeInProgrees={this.props.removeInProgrees}
        isProgress={this.props.save.isProgress}
        isError={this.props.save.isError}
      />
    );
  }
}

const mapStateToProps = state => ({
  removeInProgrees: state.remove.removeInProgrees,
});

export default connect(mapStateToProps)(SaveControlContainer);
