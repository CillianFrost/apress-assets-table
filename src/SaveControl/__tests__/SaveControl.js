import React from 'react';
import {shallow, mount, render} from 'enzyme';
import SaveControl from '../SaveControl';
import * as actions from '../actions';

describe('SaveControl', () => {
  const message = {
    success: 'Все изменения сохранены',
    progress: 'Изменения сохраняются',
    error: 'Ошибка сохранения',
  };

  it('should render a SaveControl state success', () => {
    const wrapper = shallow(
      <SaveControl
        message={message}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a SaveControl state error', () => {
    const wrapper = shallow(
      <SaveControl
        isError={true}
        message={message}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a SaveControl state progress', () => {
    const wrapper = shallow(
      <SaveControl
        isProgress={true}
        message={message}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
