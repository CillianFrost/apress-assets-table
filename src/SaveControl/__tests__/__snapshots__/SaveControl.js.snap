// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaveControl should render a SaveControl state error 1`] = `
<div
  className={[Function]}
>
  Ошибка сохранения
</div>
`;

exports[`SaveControl should render a SaveControl state progress 1`] = `
<div
  className={[Function]}
>
  Изменения сохраняются
</div>
`;

exports[`SaveControl should render a SaveControl state success 1`] = `
<div
  className={[Function]}
>
  Все изменения сохранены
</div>
`;
